Move DeviceInfoProvider to common code and use it in linux and NRF.

andy31415 opened the pull request:
#### Problem
#17596 reports that DeviceInfoProvider is not supported in nrfconnect (specifically supported locales list is empty).

This fixes #17596.

#### Change overview
Move DeviceInfoProvider implementation as an example provider and make use of it in linux and nrf.

Tried to preserve existing behavior and have linux (and darwin app main as a sideffect) support deviceinfo provider. Ameba uses a custom build script that is out of the tree, so I could not update that one).

#### Testing
Ran

```sh
#!/usr/bin/env bash

set -ex

function finish {
  echo "STATUS: $?"
}

trap finish EXIT

export NODEID=$RANDOM

export THREAD_CRED=hex:... # my test thread credentials here
export CHIP_TOOL="./out/linux-x64-chip-tool-ipv6only/chip-tool"

log_file="run_log_$(date +'%Y-%m-%d_%H-%M-%S').txt"
rm /tmp/chip_*

echo "NODE ID USED: $NODEID" | tee $log_file

echo "=============================================================" >>$log_file
echo $CHIP_TOOL pairing ble-thread $NODEID $THREAD_CRED 20202021 3840 | tee -a $log_file
$CHIP_TOOL pairing ble-thread $NODEID $THREAD_CRED 20202021 3840  >>$log_file

echo "=============================================================" >>$log_file
echo $CHIP_TOOL localizationconfiguration read supported-locales $NODEID 0 | tee -a $log_file
$CHIP_TOOL localizationconfiguration read supported-locales $NODEID 0 | tee -a $log_file | grep 'CHIP:DMG:'
```

and saw output that was non-empty:

```
...
[1653067684.692432][838470:838477] CHIP:DMG:                    AttributeDataIB =
[1653067684.692434][838470:838477] CHIP:DMG:                    {
[1653067684.692436][838470:838477] CHIP:DMG:                            DataVersion = 0x2179c2af,
[1653067684.692438][838470:838477] CHIP:DMG:                            AttributePathIB =
[1653067684.692440][838470:838477] CHIP:DMG:                            {
[1653067684.692442][838470:838477] CHIP:DMG:                                    Endpoint = 0x0,
[1653067684.692444][838470:838477] CHIP:DMG:                                    Cluster = 0x2b,
[1653067684.692447][838470:838477] CHIP:DMG:                                    Attribute = 0x0000_0001,
[1653067684.692448][838470:838477] CHIP:DMG:                                    ListIndex = Null,
[1653067684.692450][838470:838477] CHIP:DMG:                            }
[1653067684.692453][838470:838477] CHIP:DMG:
[1653067684.692456][838470:838477] CHIP:DMG:                                    Data = "ja-JP",
[1653067684.692458][838470:838477] CHIP:DMG:                    },
[1653067684.692462][838470:838477] CHIP:DMG:
[1653067684.692463][838470:838477] CHIP:DMG:            },
[1653067684.692468][838470:838477] CHIP:DMG:
[1653067684.692470][838470:838477] CHIP:DMG:    ],
[1653067684.692494][838470:838477] CHIP:DMG:
[1653067684.692497][838470:838477] CHIP:DMG:    SuppressResponse = true,
[1653067684.692499][838470:838477] CHIP:DMG:    InteractionModelRevision = 1
[1653067684.692501][838470:838477] CHIP:DMG: }
[1653067684.693219][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693221][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693222][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693223][838470:838470] CHIP:DMG: IM WH moving to [Uninitialized]
[1653067684.693225][838470:838470] CHIP:DMG: All ReadHandler-s are clean, clear GlobalDirtySet
+ finish
+ echo 'STATUS: 0'
STATUS: 0
```

bzbarsky-apple commented on src/platform/Zephyr/DeviceInfoProviderImpl.cpp:
```diff
@@ -0,0 +1,366 @@
+
+size_t DeviceInfoProviderImpl::FixedLabelIteratorImpl::Count()
+{
+    // In Zephyr Simulation, return the size of the hardcoded labelList on all endpoints.
```
What is "Zephyr Simulation"?  How would an actual device maker implement an actual fixed label cluster on top of this code?

andy31415 replied:
It is basically a copy & paste from other platforms - I see amebad and linux being basically the same thing.

As I understand it, each application needs to implement this, however the paths to bind this at application level seems to not be implemented and this exists rather at the platform level.

If we are to re-make the bindings, I believe this should happen as a separate PR/effort. Right now this is "do the same as linux/darwin/ameba".

bzbarsky-apple replied:
I know it's a copy/paste.  My point is that for Linux and Darwin this approach maybe made sense in the "this is just test code, no one is going to actually use it" sense, but in general this approach does not make sense, and checking it in just gives us a false sense of "it's implemented" when it's actually not.

Let me turn the question around: what is the point of adding this code?  As far as I can tell, the only point is to make it looks like tests pass when they would not actually do that for a real application.  I don't think we should be going there.

andy31415 replied:
Chatted offline, I will move the existing implementation (which seems to be copied and pasted across implementations) to a central place that can be reused by various app main.

andy31415 replied:
Updated to a common implementation.

bzbarsky-apple commented on src/platform/Zephyr/DeviceInfoProviderImpl.cpp:
```diff
@@ -0,0 +1,366 @@
+    DefaultStorageKeyAllocator keyAlloc;
+    uint16_t len = static_cast<uint16_t>(sizeof(val));
+
+    return mStorage->SyncGetKeyValue(keyAlloc.UserLabelLengthKey(endpoint), &val, len);
```
This allows unbounded storage.

I think all the comments at https://github.com/project-chip/connectedhomeip/pull/18236#pullrequestreview-968677986 apply to this copy/paste.....

LuDuda reviewed (changes requested):
Thank you @andy31415 for contribution.

I agree with Boris, that this should not be implemented in the platform layer.

Providing example implementation of DeviceInfoProvider for all platforms seems to be better solution (similarly what we do in attestation credentials `SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());`)

andy31415 commented:
> Thank you @andy31415 for contribution.
> 
> I agree with Boris, that this should not be implemented in the platform layer.
> 
> Providing example implementation of DeviceInfoProvider for all platforms seems to be better solution (similarly what we do in attestation credentials `SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());`)

@LuDuda I just updated the PR and moved things to a common layer.

NRF uses a custom cmake build which required a bit of updating. Please take a look.

Damian-Nordic commented on examples/providers/BUILD.gn:
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2021 Project CHIP Authors
```
2022
